pcap-dpdk.c

```diff
@@ -0,0 +1,808 @@
+
+		// init one RX queue
+		rxq_conf = dev_info.default_rxconf;
+		rxq_conf.offloads = local_port_conf.rxmode.offloads;
+		ret = rte_eth_rx_queue_setup(portid, 0, nb_rxd,
+					     rte_eth_dev_socket_id(portid),
+					     &rxq_conf,
+					     pd->pktmbuf_pool);
+		if (ret < 0)
+		{
+			pcap_fmt_errmsg_for_errno(p->errbuf, PCAP_ERRBUF_SIZE,
+			    errno, "dpdk error: rte_eth_rx_queue_setup:err=%d, port=%u",
+			    ret, portid);
+			ret = PCAP_ERROR;
+			break;	
+		}
+
+		// init one TX queue 
+		txq_conf = dev_info.default_txconf;
+		txq_conf.offloads = local_port_conf.txmode.offloads;
+		ret = rte_eth_tx_queue_setup(portid, 0, nb_txd,
+				rte_eth_dev_socket_id(portid),
+				&txq_conf);
+		if (ret < 0)
+		{
+			pcap_fmt_errmsg_for_errno(p->errbuf, PCAP_ERRBUF_SIZE,
+			    errno, "dpdk error: rte_eth_tx_queue_setup:err=%d, port=%u",
+			    ret, portid);
+			ret = PCAP_ERROR;
+			break;	
+		}
+		// Initialize TX buffers 
+		tx_buffer = rte_zmalloc_socket(DPDK_TX_BUF_NAME,
+				RTE_ETH_TX_BUFFER_SIZE(MAX_PKT_BURST), 0,
+				rte_eth_dev_socket_id(portid));
+		if (tx_buffer == NULL)
+		{
+			pcap_fmt_errmsg_for_errno(p->errbuf, PCAP_ERRBUF_SIZE,
+			    errno, "dpdk error: Cannot allocate buffer for tx on port %u", portid);
+			ret = PCAP_ERROR;
+			break;	
+		}
+		rte_eth_tx_buffer_init(tx_buffer, MAX_PKT_BURST);
+		// Start device
+		ret = rte_eth_dev_start(portid);
+		if (ret < 0)
+		{
+			pcap_fmt_errmsg_for_errno(p->errbuf, PCAP_ERRBUF_SIZE,
+			    errno, "dpdk error: rte_eth_dev_start:err=%d, port=%u",
+			    ret, portid);
+			ret = PCAP_ERROR;
+			break;
+		}
+		// set promiscuous mode
+		if (p->opt.promisc){
+			pd->must_clear_promisc=1;
+			rte_eth_promiscuous_enable(portid);
+		}
+		// check link status
+		is_port_up = check_link_status(portid, &link);
+		if (!is_port_up){
+			pcap_fmt_errmsg_for_errno(p->errbuf, PCAP_ERRBUF_SIZE,
+			    errno, "dpdk error: link is down, port=%u",portid);
+			ret = PCAP_ERROR_IFACE_NOT_UP;
+			break;
+		}
+		// reset statistics
+		rte_eth_stats_reset(pd->portid);
+		calculate_timestamp(&(pd->ts_helper), &(pd->prev_ts));
+		rte_eth_stats_get(pd->portid,&(pd->prev_stats));	
+		// format pcap_t 
+		pd->portid = portid;
+		p->fd = pd->portid; 
+		if (p->snapshot <=0 || p->snapshot> MAXIMUM_SNAPLEN)
+		{
+			p->snapshot = MAXIMUM_SNAPLEN;
+		}
+		p->linktype = DLT_EN10MB; // Ethernet, the 10MB is historical.
+		p->selectable_fd = p->fd;
+		p->read_op = pcap_dpdk_dispatch;
+		p->inject_op = pcap_dpdk_inject;
+		// using pcap_filter currently, though DPDK provides their own BPF function. Because DPDK BPF needs load a ELF file as a filter.
+		p->setfilter_op = install_bpf_program;
+		p->setdirection_op = NULL;
+		p->set_datalink_op = NULL;
+		p->getnonblock_op = pcap_dpdk_getnonblock;
+		p->setnonblock_op = pcap_dpdk_setnonblock;
+		p->stats_op = pcap_dpdk_stats;
+		p->cleanup_op = pcap_dpdk_close;
+		p->breakloop_op = pcap_breakloop_common;
```

Comment:
Here's where you set, and use, the `required_select_timeout` member of the private data structure.

Do

    pd->required_select_timeout.tv_sec = 0;
    pd->required_select_timeout.tv_usec = DPDK_DEF_MIN_SLEEP_MS*1000;
    p->required_select_timeout = &pd->required_select_timeout;

This way, the `pcap_get_required_select_timeout()` call will return that timeout, so that if code using pcap has a `select()`/`poll()`/`epoll()`/etc. loop, waiting for, among other input sources, a `pcap_t`, it can use that timeout to poll a DPDK `pcap_t`.  (This mechanism was introduced for DAG cards, but DPDK, like DAG, works by polling, rather than with interrupts and wakeups.)

Reply:
OK. Thanks.